Add CIS Benchmark documentation to KubeOne

koksay opened the pull request:
This PR adds CIS Benchmark 1.8 documentation for KubeOne 1.7.3 and Kubernetes 1.27.

It especially requires a review for:
1. Formatting in general
2. `Details` section for `Not Applicable` controls
3. `Warn` and `Fail` controls, if they should be changed to `Not Applicable`

embik commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,569 @@
+---
+#### 4.2.9: Ensure that the --tls-cert-file and --tls-private-key-file arguments are set as appropriate (Manual)
+
+**Result:** 🔵 Not Applicable
+
+**Details:** This is a manual check, `--tls-cert-file` and `--tls-private-key-file` options are provided to Kubelet
```
I think this should be Pass. We can still keep the details.

koksay replied:
Ok, `not applicable` because of my script generation. I will update change it to `pass`

embik commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,569 @@
+
+**Result:** 🔵 Not Applicable
+
+**Details:** DenyServiceExternalIPs is not enabled by default, and it is not in the enabled plugins list
```
As a user, I don't understand why this is not applicable. Can we please elaborate what this is and why we don't think it's needed?

embik commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,569 @@
+### 1.2. API Server
+#### 1.2.1: Ensure that the --anonymous-auth argument is set to false (Manual)
+
+**Result:** 🔵 Not Applicable
+
+**Details:** This is mitigated by RBAC, please see https://kubernetes.io/docs/reference/access-authn-authz/authentication/#anonymous-requests
```
I'm not sure this is sufficient. At best, we'd disable anonymous auth completely. Do we have any idea why this isn't set by KubeOne/kubeadm?

embik commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,569 @@
+---
+#### 1.1.9: Ensure that the Container Network Interface file permissions are set to 600 or more restrictive (Manual)
+
+**Result:** 🟡 Warn
+
+_The issue is under investigation to provide a fix in a future KubeOne release_
```
What's the difference between Warn and Fail? Should we maybe explain that the beginning of the document?

koksay replied:
Actually, this is explained on top of the document, but I can update the wording.

embik replied:
No, I just missed it for some reason. Sorry 🤦🏻

koksay commented:
@embik I believe I covered all the items except `--anonymous-auth` parameter setting. To be honest, I have no idea why we don't set it at all (other than it does not cause issues because or our RBAC settings)

embik commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,569 @@
+
+**Result:** 🟠 Configurable
+
+**Details:** Audit logging is not enabled by default, it can be configured as described here: https://docs.kubermatic.com/kubeone/v1.7/tutorials/creating-clusters-oidc/#audit-logging
```
The URLs should be `ref` instead of absolute URL, so they link to the right version of the documentation. Here's an example: https://github.com/kubermatic/docs/blob/master/content/kubeone/main/_index.en.md?plain=1#L78

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,574 @@
++++
+title = "Benchmark on KubeOne-1.7.3 and Kubernetes-1.27"
```
We don't use hyphen to separate product name and version. Also, I'd consider rotating the Kubernetes version and the KubeOne version, and dropping the KubeOne patch release:
```suggestion
title = "Benchmark on Kubernetes 1.27 with KubeOne 1.7"
```

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,574 @@
+date = 2024-03-06T12:01:00+02:00
++++
+
+This benchmark guide helps you evaluate the security of a KubeOne cluster against each control in the CIS Kubernetes Benchmark.
```
Some minor nits:
```suggestion
This guide helps you evaluate the security of a Kubernetes cluster created using KubeOne against each control in the CIS Kubernetes Benchmark.
```

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,574 @@
+
+| KubeOne Version  | Kubernetes Version | CIS Benchmark Version |
+| ---------------- | ------------------ | --------------------- |
+| 1.7.3               | 1.27                 | CIS-1.8                    |
```
Nit: please add the Kubernetes patch release that you used

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,574 @@
+
+These are the possible results for each control:
+
+🟢 **Pass:** The KubeOne cluster passes the audit outlined in the benchmark.
```
```suggestion
🟢 **Pass:** The KubeOne cluster passes the audit/control outlined in the benchmark.
```

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,574 @@
+
+🔵 **Pass (Additional Configuration Required):** The KubeOne cluster passes the audit outlined in the benchmark with some extra configuration. The documentation is provided.
+
+🔴 **Fail:** The control will be fixed in a future KubeOne release.
```
```suggestion
🔴 **Fail:** The audit/control will be fixed in a future KubeOne release.
```

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,574 @@
+**Result:** 🟢 Pass
+
+---
+#### 1.2.9: Ensure that the admission control plugin EventRateLimit is set (Manual)
+
+**Result:** 🟢 Pass
+
+**Details:** EventRateLimit admission control plugin in in Alpha state, please see [here][eventratelimit]
```
I was thinking more about this audit and I think it's actually a fail. We don't enable this admission control plugin, CIS requires it and doesn't go into details if it's alpha or not. It's up to us to decide if we want to support an alpha feature or not (we have always been towards not supporting alpha features and we can check with @csengerszabo if it's safe to state so in this document)

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/cis1.8-kubeone1.7-k8s1.27/_index.en.md:
```diff
@@ -0,0 +1,574 @@
+_The issue is under investigation to provide a fix in a future KubeOne release_
+
+---
+#### 1.2.12: Ensure that the admission control plugin SecurityContextDeny is set if PodSecurityPolicy is not used (Manual)
+
+**Result:** 🟢 Pass
+
+**Details:** SecurityContextDeny admission control plugin is deprecated as of [Kubernetes 1.27][securitycontextdeny]
```
When I think better, this is also a fail, for the same reason as the previous comment. We don't enable this admission control plugin and that's all what CIS cares about. We can say that this audit failed but that we don't support enabling this admission control plugin because the official Kubernetes docs state that it shouldn't be used at all and that it's deprecated.

xmudrii reviewed (changes requested):
The term "KubeOne cluster" is used a lot through both documents but I'm not a huge fan of that term. I much more prefer using the term "Kubernetes cluster created using KubeOne" or something similar. "KubeOne cluster" sounds like a cluster of KubeOnes, but in reality it is a cluster of Kubernetes nodes. Other than that and other comments I left, this looks great to me!

koksay commented:
@xmudrii Thanks again for the review. I updated the document based on your feedback.

I also added my doc generation python script to the `hack` directory, so it would be used later.

xmudrii commented on content/kubeone/main/guides/cis-benchmarking/_index.en.md:
```diff
@@ -0,0 +1,42 @@
+
+[CIS Benchmark for Kubernetes](https://www.cisecurity.org/benchmark/kubernetes) is a guide that consists of secure configuration guidelines and best practices developed for Kubernetes.
+
+In this document, information how it can be run on KubeOne clusters and what to expect as the result is described.
```
Nit:
```suggestion
In this document, information how it can be run on a Kubernetes cluster created using KubeOne and what to expect as the result is described.
```

xmudrii commented on hack/kube-bench-k1/master.json:
Can we gitignore this file?

xmudrii commented on hack/kube-bench-k1/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## Running kube-bench
+
+Kube bench is deployed on the master node, and run this way:
```
Nit:
```suggestion
kube-bench is deployed on the master node, and run this way:
```

xmudrii reviewed (changes requested):
Some minor nits, after those are fixed, we can merge this PR and iterate later on if needed

koksay commented:
@xmudrii I put it on the `main` documentation, should it also be under `1.7`?

xmudrii commented:
@koksay Adding it to 1.7 makes sense, do you want to do that before we merge this PR?

xmudrii reviewed (approved):
/lgtm
/approve
